### PR TITLE
feat: @rehearsal/diagnostic package

### PR DESCRIPTION
**Pull request by egorio**
Implement diagnostic package (including reporter).

- **diagnose.ts** exports a diagnose function
- **cli.ts** provides CLI interface (will be moved to @rehearsal/cli after @rehearsal/diagnostic is published)

### CODE REVIEW

**Review by lynchbomb — CHANGES_REQUESTED** [review]

**Comment by lynchbomb** [comment]
Left a bunch of comments. Critically we dont need the custom reporter or cli within the diagnostic package as these are over complicating things and duplicating work. instead we should extend what we already have (reporter and cli packages). 

Just export a module for the diagnostic and if there's a real need for it we can create a command within the existing cli for `diagnostic` etc. But for now keep it only as a module export (no cli).

WRT the reporter, just remove the custom reporter in the diagnostic package and import the reporter from the rehearsal workspace. Change/Add an interface of the diagnostic implementation delta into the reporter, extending its functionality as needed (in a new PR).

File: packages/diagnostic/test/fixtures/diagnose/.gitignore [hunk 1]
```diff
@@ -0,0 +1,6 @@
+first.ts
```
[the remark this answers is not in the document]
Comment (egorio) [reply packages/diagnostic/test/fixtures/diagnose/.gitignore:1]: This directory `fixtures/diagnose` represents an independent app for the test, this is why it has it's own configs and .gitignore.

File: packages/diagnostic/src/cli.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
+import commander from 'commander';
```
[the remark this answers is not in the document]
Comment (egorio) [reply packages/diagnostic/src/cli.ts:1]: Agree. Let's move this command to `/cli` after `@rehearsal/diagnostic` package will be published to `npm`

File: packages/diagnostic/package.json [hunk 3]
```diff
@@ -28,14 +29,18 @@
     "test:coverage": "yarn nyc mocha"
   },
   "dependencies": {
```
[the remark this answers is not in the document]
Comment (egorio) [reply packages/diagnostic/package.json:31]: answered above

File: packages/diagnostic/bin/diagnose.sh [hunk 4]
```diff
@@ -0,0 +1,3 @@
+#!/usr/bin/env bash
+
+node "$(dirname $0)/../dist/src/cli.js" "$@"
```
[the remark this answers is not in the document]
Comment (egorio) [reply packages/diagnostic/bin/diagnose.sh:3]: answered above

File: packages/diagnostic/src/diagnose.ts [hunk 5]
```diff
@@ -1,76 +1,176 @@
...
+      text = service.printSourceFile(service.getSourceFile(fileName));
 
-      await lint(sourceFile);
-      service.updateSourceFile(sourceFile);
+      // Restore empty lines and reformat the code again
+      text = await restoreEmptyLines(text);
+      text = await fixReactTemplateComment(text, fileName);
+      text = await lint(text, fileName);
+
+      await fs.promises.writeFile(fileName, text);
```
[the remark this answers is not in the document]
Comment (egorio) [reply packages/diagnostic/src/diagnose.ts:100]: No reason, except everything called async there :)
The `fs` has `writeFileSync` by itself.

File: packages/diagnostic/src/diagnostic-reporter.ts [hunk 6]
```diff
@@ -0,0 +1,54 @@
+import fs from 'fs';
```
[the remark this answers is not in the document]
Comment (egorio) [reply packages/diagnostic/src/diagnostic-reporter.ts:1]: I expect to move this reporter to `@rehearsal/reporter` module when will be working on `@rehersal/migration` and use it as the connection between to those modules.

File: packages/diagnostic/src/diagnose.ts [hunk 7]
```diff
@@ -1,76 +1,176 @@
+import fs from 'fs';
```
[the remark this answers is not in the document]
Comment (egorio) [reply packages/diagnostic/src/diagnose.ts:1]: I expect to move this reporter to @rehearsal/reporter module when will be working on @rehersal/migration and use it as the connection between to those modules.

**Comment by egorio** [comment]
> Left a bunch of comments. Critically we don't need the custom reporter or cli within the diagnostic package as these are over complicating things and duplicating work. instead we should extend what we already have (reporter and cli packages).

This is exact the plan.

I'm going to refactor `@rehearsal/reporter` to keep it simple and responsible for writing / reading (need for migration) report file. CLI command will be migrated to `@reporter/cli` once we publish the package and I will be able to add it as dependency.

**Pull request merged**